stl/inc/xutility

```diff
@@ -4889,23 +4889,120 @@ public:
+        const uint64_t __a, const uint64_t __b, uint64_t* const __productHi) {
+        return _umul128(__a, __b, __productHi);
+    }
+
+    _NODISCARD static inline uint64_t __rng_shiftright128(
+        const uint64_t __lo, const uint64_t __hi, const uint32_t __dist) {
+        // For the __shiftright128 intrinsic, the shift value is always
+        // modulo 64.
+        // In the current implementation of the double-precision version
+        // of Ryu, the shift value is always < 64.
+        // (The shift value is in the range [49, 58].)
+        // Check this here in case a future change requires larger shift
+        // values. In this case this function needs to be adjusted.
+        _STL_INTERNAL_CHECK(__dist < 64);
+        return __shiftright128(__lo, __hi, static_cast<unsigned char>(__dist));
+    }
+
+#else // ^^^ intrinsics available ^^^ / vvv intrinsics unavailable vvv
+
+    _NODISCARD static __forceinline uint64_t __rng_umul128(
+        const uint64_t __a, const uint64_t __b, uint64_t* const __productHi) {
+        // TRANSITION, VSO-634761
+        // The casts here help MSVC to avoid calls to the __allmul library function.
+        const uint32_t __aLo = static_cast<uint32_t>(__a);
+        const uint32_t __aHi = static_cast<uint32_t>(__a >> 32);
+        const uint32_t __bLo = static_cast<uint32_t>(__b);
+        const uint32_t __bHi = static_cast<uint32_t>(__b >> 32);
+
+        const uint64_t __b00 = static_cast<uint64_t>(__aLo) * __bLo;
+        const uint64_t __b01 = static_cast<uint64_t>(__aLo) * __bHi;
+        const uint64_t __b10 = static_cast<uint64_t>(__aHi) * __bLo;
+        const uint64_t __b11 = static_cast<uint64_t>(__aHi) * __bHi;
+
+        const uint32_t __b00Lo = static_cast<uint32_t>(__b00);
+        const uint32_t __b00Hi = static_cast<uint32_t>(__b00 >> 32);
+
+        const uint64_t __mid1   = __b10 + __b00Hi;
+        const uint32_t __mid1Lo = static_cast<uint32_t>(__mid1);
+        const uint32_t __mid1Hi = static_cast<uint32_t>(__mid1 >> 32);
+
+        const uint64_t __mid2   = __b01 + __mid1Lo;
+        const uint32_t __mid2Lo = static_cast<uint32_t>(__mid2);
+        const uint32_t __mid2Hi = static_cast<uint32_t>(__mid2 >> 32);
+
+        const uint64_t __pHi = __b11 + __mid1Hi + __mid2Hi;
+        const uint64_t __pLo = (static_cast<uint64_t>(__mid2Lo) << 32) | __b00Lo;
+
+        *__productHi = __pHi;
+        return __pLo;
+    }
+
+    _NODISCARD static inline uint64_t __rng_shiftright128(
+        const uint64_t __lo, const uint64_t __hi, const uint32_t __dist) {
+        // We don't need to handle the case __dist >= 64 here (see above).
+        _STL_INTERNAL_CHECK(__dist < 64);
+#ifdef _WIN64
+        _STL_INTERNAL_CHECK(__dist > 0);
+        return (__hi << (64 - __dist)) | (__lo >> __dist);
+#else // ^^^ 64-bit ^^^ / vvv 32-bit vvv
+      // Avoid a 64-bit shift by taking advantage of the range of shift values.
+        _STL_INTERNAL_CHECK(__dist >= 32);

```

Comment:
This code was copied from charconv/Ryu (as previous comments indicate). This raises a couple of issues which need to be addressed: (1) as this comment indicates, there is a Ryu-specific assumption about the range of possible shift values which is probably not applicable here, and (2) if we copy Ryu code around in our repo, we need to ensure that the Boost license banner with Ulf Adams' copyright notice appears in those files.

(This also means that you should test both x86 and x64.)

Reply:
Yes, I missed the range restriction in the x86 case. That need's to be adapted.

Regarding (2), I'm not sure how to correctly achieve this. Would it "just" be adding the copyright section of xcharconv_ryu.h (lines 7-32) to xutility? Perhaps as an alternative, I'd ask if copying the code is a good idea in the first place or if a separate header providing a ´_Umul128´ function is preferable, which would also make attribution more clear.

Reply:
I'm increasingly liking your idea of a separate header for the attribution reason you've mentioned. We generally try to avoid having lots of tiny headers, but it's not an ironclad rule, and we could even avoid code duplication (since the umul128 doesn't need to be modified).

For now, what you have is fine. I'll need to think about how to rearrange this with minimal disruption to charconv (where https://github.com/StephanTLavavej/ryu/tree/msvc-latest is layering a massive number of changes on top of the upstream sources); when I figure that out, I'll want to push the change to your branch. I'll leave this active as a reminder to myself.